### PR TITLE
Export Filename.{Unix,Win32,Cygwin}

### PR DESCRIPTION
**Pull request by xvw**
Export submodules Unix, Win32 and Cygwin.

### CODE REVIEW

**Comment by dra27** [comment]
I believe the original intent of this was also to expose `has_drive`. See, for example, https://gitlab.com/camlspotter/ocamlspot/-/blob/hg-default/filepath.ml#L40-46.

I spent some time on this a week or so ago thinking it would be quick and found quite a few bugs (especially in the Cygwin handling), which I'll push. The only problem with this as an interim step is that I don't think it actually fixes the original report's issue.

**Comment by xvw** [comment]
@dra27 Yes you are right. I'm not very familar with "has_drive". So I've just remove the reference of "fix". Is that ok for you?

**Comment by nojb** [comment]
Does `SYSDEPS` coincide with the full signature of the (current) `Filename` module? If yes, wouldn't it be better to `include SYSDEPS` in the interface of `Filename`?

**Comment by xvw** [comment]
@nojb There is some small differences (`concat` is not in `SYSDEP` and `is_dir_sep` is not in the full API). The code could be reduced using a Functor but my goal in this PR was "only" to expose submodules. Do you think that I should (in this PR) go through a more ambitious refactoring?

**Comment by nojb** [comment]
> @nojb There is some small differences (`concat` is not in `SYSDEP` and `is_dir_sep` is not in the full API). The code could be reduced using a Functor but my goal in this PR was "only" to expose submodules. Do you think that I should (in this PR) go through a more ambitious refactoring?

The problem is that we are stuck with any change in the `stdlib` more or less "forever", so we need to be careful. Personally, I think that if the objective of the PR is to simultaneously expose a version of `Filename` for each supported system, then using a common signature for the submodules and the "main" `Filename` module makes sense. But I haven't looked at the implementation and have no idea how difficult this is.

**Comment by xvw** [comment]
@nojb Is it more like you thought about it?

File: otherlibs/unix/unix.mli [hunk 1]
```diff
@@ -237,7 +237,7 @@ val system : string -> process_status
    therefore can contain redirections, quotes, variables, etc.
    To properly quote whitespace and shell special characters occurring
    in file names or command arguments, the use of
-   {!Filename.quote_command} is recommended.
+   {!Filename.FILENAME.quote_command} is recommended.
```
Comment (nojb) [inline otherlibs/unix/unix.mli]: The previous docstring is better, as 1) there is no value `Filename.FILENAME.quote_command`, and 2) the function that should be used is the one corresponding to the current system, so it is just `Filename.quote_command`.

File: otherlibs/unix/unix.mli [hunk 2]
```diff
@@ -784,7 +784,7 @@ val open_process_in : string -> in_channel
    which can be read via the returned input channel.
    The command is interpreted by the shell [/bin/sh]
    (or [cmd.exe] on Windows), cf. {!Unix.system}.
-   The {!Filename.quote_command} function can be used to
+   The {!Filename.FILENAME.quote_command} function can be used to
```
Comment (nojb) [inline otherlibs/unix/unix.mli]: Idem.

File: stdlib/filename.ml [hunk 3]
```diff
@@ -13,6 +13,58 @@
...
+  val open_temp_file :
+    ?mode: open_flag list -> ?perms: int -> ?temp_dir: string -> string ->
+    string -> string * out_channel
+
+  val get_temp_dir_name : unit -> string
+
+  val set_temp_dir_name : string -> unit
+
+  val temp_dir_name : string
+  [@@ocaml.deprecated "Use Filename.get_temp_dir_name instead"]
```
Comment (nojb) [inline stdlib/filename.ml]: As this one is deprecated, would it be difficult to only include it in the top-level module (for backwards-compatibility), but not in the new submodules?

File: stdlib/filename.ml [hunk 4]
```diff
@@ -13,6 +13,58 @@
 (*                                                                        *)
 (**************************************************************************)
 
+module type FILENAME = sig
```
Comment (nojb) [inline stdlib/filename.ml]: Should we call this signature `S`, by analogy to other existing modules in the standard library (`Hashtbl.S`, `Map.S`, ...)?

File: stdlib/filename.ml [hunk 5]
```diff
@@ -261,99 +389,33 @@ Quoting commands for execution by cmd.exe is difficult.
...
 end
 
-module Sysdeps =
+module Unix = Make (Unix_deps)
+
+module Win32 = Make (Win32_deps)
+
+module Cygwin = Make (Cygwin_deps)
+
+module Current =
```
Comment (nojb) [inline stdlib/filename.ml]: Not very important, but you can just write `include` here instead of `module Current =`.

File: stdlib/sys.mli [hunk 6]
```diff
@@ -82,7 +82,7 @@ external command : string -> int = "caml_sys_system_command"
   command names or in their arguments must be quoted or escaped
   so that the shell does not interpret them.  The quoting rules vary
   between the POSIX shell and the Windows shell.
-  The {!Filename.quote_command} performs the appropriate quoting
+  The {!val:Filename.FILENAME.quote_command} performs the appropriate quoting
```
Comment (nojb) [inline stdlib/sys.mli]: As with `Unix`, here the previous docstring is better.

File: stdlib/filename.mli [hunk 7]
```diff
@@ -83,27 +90,29 @@ val extension : string -> string
...
     @since 4.04
-*)
+   *)
 
-val remove_extension : string -> string
-(** Return the given file name without its extension, as defined
-    in {!Filename.extension}. If the extension is empty, the function
+  val remove_extension : string -> string
+  (** Return the given file name without its extension, as defined
+    in {!val:extension}. If the extension is empty, the function
```
Comment (nojb) [inline stdlib/filename.mli:93]: Have you checked that the generated `ocamldoc` looks OK?

File: stdlib/filename.ml [hunk 8]
```diff
@@ -83,51 +135,127 @@ module type SYSDEPS = sig
...
+    else String.sub name 0 (String.length name - l)
+
+  let remove_extension name =
+    let l = extension_len name in
+    if l = 0 then name else String.sub name 0 (String.length name - l)
+
+  external open_desc: string -> open_flag list -> int -> int = "caml_sys_open"
+  external close_desc: int -> unit = "caml_sys_close"
+
+  let prng = lazy(Random.State.make_self_init ())
```
Comment (nojb) [inline stdlib/filename.ml]: This (and the two external above) could be put outside of the functor, I think.

**Review by nojb — CHANGES_REQUESTED** [review]
Thanks, this is what I had in mind. I left some suggestions, but in principle this looks good to me.

As this is a stdlib change, it would be better to have a second opinion a well. Perhaps @dra27 or @gasche can take a look?

File: otherlibs/unix/unix.mli (hunk 1 above)
Comment (xvw) [reply otherlibs/unix/unix.mli]: It does not work (a warning is emitted) with OCamldoc.

File: stdlib/filename.mli (hunk 7 above)
Comment (xvw) [reply stdlib/filename.mli:93]: Yes (it refer Filename.FILENAME.extension).

File: stdlib/filename.ml (hunk 3 above)
Comment (xvw) [reply stdlib/filename.ml]: Done

File: stdlib/filename.ml (hunk 4 above)
Comment (xvw) [reply stdlib/filename.ml]: Yes, you're right. Done.

File: stdlib/filename.ml (hunk 5 above)
Comment (xvw) [reply stdlib/filename.ml]: Done!

File: stdlib/filename.ml (hunk 8 above)
Comment (xvw) [reply stdlib/filename.ml]: You're right, I've put the externals and prng outside of the functor. Thanks.

**Comment by xvw** [comment]
There is still (in my point of view) a discussion point. Ocamldoc does not handle (correctly) inclusion (#7332). It is why I have to reference `Filename.S.f` instead of `Filename.f` and `S` is not expanded in the documentation of Filename, which can be confusing. Should I duplicate like this: 

```ocaml
module type S = sig 
    val f : a -> b 
    (** doc **)
end

val f : a -> b
(** doc **)
```

Instead of: 

```ocaml
module type S = sig 
    val f : a -> b 
    (** doc **)
end

include S
```

?

**Comment by nojb** [comment]
> There is still (in my point of view) a discussion point. Ocamldoc does not handle (correctly) inclusion (#7332). It is why I have to reference `Filename.S.f` instead of `Filename.f` and `S` is not expanded in the documentation of Filename, which can be confusing.

Not sure how hard it would be, but I think it may be worth trying to generate the docs using `odoc` as well. This is because as I understand it, the plan is to switch to `odoc` for documentation in the not-too-distant future. cc @Octachron which will have more to say about this.

**Comment by xvw** [comment]
Thanks @nojb, I have asked to @Octachron and he thought me that duplication is an acceptable compromise (until ODoc).
So I have duplicate Filename.S documentation in Filename (and removed "since" in Filename.S).

**Comment by dbuenzli** [comment]
Could someone explain how these modules are supposed to be used by end users and/or what use case they intend to solve ? 

FWIW I had a similar PR recently on my `Fpath` module and I think that simply exposing these interfaces is really not the right way to go on about the problem. 

I think it may be worth pondering the comment I made [there](https://github.com/dbuenzli/fpath/pull/13#issuecomment-645938375). One thing that is obviously different in `Filename` is that it has no abstract type but the problem (and answers I try to give there) remains the same. 

It all starts with asking you the question how are you going to know which kind of path your are handling and thus which of the modules you have to use ?

**Comment by nojb** [comment]
> It all starts with asking you the question how are you going to know which kind of path your are handling and thus which of the modules you have to use ?

Thanks for chiming in, and asking a very good question. Just to clarify my position, I am not convinced the PR should be merged, but *if* we were to merge it, then my point was that the submodules should have the same signature as the existing `Filename` module.

Having said that, if I wanted to argue for inclusion, I would put forward the case of working in mixed environments, for example, when you need to call Cygwin executables from a native Windows port (or the other way around), or calling native Windows programs from inside WSL port. The user is assumed to know the expected encoding of the paths she is building in each case. The point of the PR is to make it possible, *under this assumption*, to use the right function to, say, quote a program argument.

But I agree that the exposed API is not a full solution to the problem of handling and mixing paths between different systems. Personally I am not aware of any satisfactory API to solve this problem in other languages, but would love to hear otherwise.

Given all this, the question seems to be: do we want to go ahead and expose these submodules for the convenience in those few situations where one knows exactly the conventions of every path in sight, or do we consider that exposing these submodules is akin to exposing a kind of "unsafe" API that will do more harm than good?

In my experience, the opportunities where I would have made use of these submodules where few and far between, so I think it would be perfectly reasonable to wait instead until a better API comes along.

**Comment by xvw** [comment]
@Et7f3 yes, I think it is fixed on the last commit?

My goal with this PR was to solve #6287. But it is maybe not very useful so should I close the PR?

**Comment by Et7f3** [comment]
Ah sorry I haven't seen this PR: just see my commit cherry-pick and notified space. Yes it is fixed. Sorry for the noise. (btw their is also some whitespace change, but I don't know the policy of this repo about whitespace change)

> so should I close the PR?

No!! This is a useful PR.

**Comment by xvw** [comment]
I think whitspace change are present because I move the content of the module in a submodule.

**Comment by nojb** [comment]
> My goal with this PR was to solve #6287. But it is maybe not very useful so should I close the PR?

I would suggest to give some time for other developers to chime in before taking any decision. Thanks!

**Comment by hcarty** [comment]
> the case of working in mixed environments, for example, when you need to call Cygwin executables from a native Windows port (or the other way around), or calling native Windows programs from inside WSL port. The user is assumed to know the expected encoding of the paths she is building in each case. The point of the PR is to make it possible, under this assumption, to use the right function to, say, quote a program argument.

This is a case I run into in the real world. While it doesn't happen frequently it is very challenging to deal with given the current stdlib and OCaml library ecosystem.

I understand @dbuenzli's concerns about this exposing an incomplete API.  However, without this a lot of the useful tools to build a more complete API exist in OCaml but are hidden from users.  Exposing these platform-specific modules makes it easier for the community to experiment with higher level logic built on top of `Filename` and addresses the case(s) @nojb mentioned where a user knows the environment a path is targeting.

I would argue for including this PR because it unblocks a real world use case without introducing lots of new code and it makes it easier for users to build more sophisticated APIs if those APIs are deemed useful.

**Comment by dbuenzli** [comment]
> I understand @dbuenzli's concerns about this exposing an incomplete API.

My concern is not about exposing an incomplete API it's about exposing a wrong model that the stdlib will have to support for ages.

> I would argue for including this PR because it unblocks a real world use case without introducing lots of new code and it makes it easier for users to build more sophisticated APIs if those APIs are deemed useful.

As far as I can see, the stdlib evolution tends to work the other way round, once it is sure something works and is needed somewhat pervasively, or solves a library interoperability problem, the case is made for introduction. It's perfectly possible to experiment with this outside the stdlib without this PR, even more so as `Filename` exposes no abstract type.

Someone has to come up with a design and experiment with it, bulk exposing library internals is not going to magically solve the problem. I have sketched one in the `fpath` PR I mentioned earlier, I currently don't have the need for it and no money to invest to develop it myself. But I'm happy to integrate the work in `fpath` from anyone interested in following the ideas outlined there.

**Comment by xvw** [comment]
Since the API is incomplete. I suggest to close the PR.
Thanks for your reviews and your insights!

**Pull request closed without merging**